archaius2-core/src/main/java/com/netflix/archaius/config/PollingDynamicConfig.java

```diff
@@ -127,6 +142,42 @@ public Iterable<String> keys() {
 
     @Override
     public void forEachProperty(BiConsumer<String, Object> consumer) {
+        boolean instrumentationEnabled = instrumentationEnabled();
+        current.forEach((k, v) -> {
+            if (instrumentationEnabled) {
+                recordUsage(new PropertyDetails(k, currentIds.get(k), v));
+            }
+            consumer.accept(k, v);
+        });
+    }
+
+    @Override
+    public void forEachPropertyUninstrumented(BiConsumer<String, Object> consumer) {
         current.forEach(consumer);
     }
+
+    public void setAccessMonitorUtil(AccessMonitorUtil accessMonitorUtil) {
```

Comment:
Race condition alert!
I assume you kept these two fields separate in order to allow runtime disabling/enabling of instrumentation, right? (if not, just change `instrumentationEnabled()` to just return `accessMonitorUtil != null` and skip the rest of this comment :-) )

These two MUST be set together, or you could end up getting a null pointer in line 175. In order to ensure that without synchronizing everywhere, bundle both fields into a small POJO that's kept in a volatile field. Something like (in record syntax because I'm lazy but we'll need to do it with a normal class b/c java8 :-( )
```
   private static record Instrumentation(AccessMonitorUtil monitor, boolean enabled);
   private volatile Instrumentation instrumentation;

   [...]

  public void setAccessMonitorUtil(monitor) {
     this.instrumentation = new Instrumentation(monitor, true);
  }
```

Reply:
Ah right, good catch! I updated to use an Instrumentation class to encapsulate both and added a comment for context (also moved the AccessMonitorUtil into the constructor).